{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ad7e64c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found existing installation: scikit-learn 1.1.1\n",
      "Uninstalling scikit-learn-1.1.1:\n",
      "  Successfully uninstalled scikit-learn-1.1.1\n",
      "Collecting scikit-learn==0.24.1\n",
      "  Downloading scikit_learn-0.24.1-cp39-cp39-macosx_10_13_x86_64.whl (7.3 MB)\n",
      "\u001b[K     |████████████████████████████████| 7.3 MB 9.8 MB/s eta 0:00:01\n",
      "\u001b[?25hRequirement already satisfied: threadpoolctl>=2.0.0 in /Users/hyewonhan/miniconda3/lib/python3.9/site-packages (from scikit-learn==0.24.1) (2.2.0)\n",
      "Requirement already satisfied: numpy>=1.13.3 in /Users/hyewonhan/miniconda3/lib/python3.9/site-packages (from scikit-learn==0.24.1) (1.22.3)\n",
      "Requirement already satisfied: joblib>=0.11 in /Users/hyewonhan/miniconda3/lib/python3.9/site-packages (from scikit-learn==0.24.1) (1.1.0)\n",
      "Requirement already satisfied: scipy>=0.19.1 in /Users/hyewonhan/miniconda3/lib/python3.9/site-packages (from scikit-learn==0.24.1) (1.7.3)\n",
      "Installing collected packages: scikit-learn\n",
      "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "imbalanced-learn 0.9.1 requires scikit-learn>=1.1.0, but you have scikit-learn 0.24.1 which is incompatible.\u001b[0m\n",
      "Successfully installed scikit-learn-0.24.1\n"
     ]
    }
   ],
   "source": [
    "!pip uninstall scikit-learn -y\n",
    "!pip install scikit-learn==0.24.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "00bf2097",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0ddc39ad",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>OSA</th>\n",
       "      <th>insomnia</th>\n",
       "      <th>PLMD, RLS</th>\n",
       "      <th>parasomnia</th>\n",
       "      <th>circadian rhythm sleep-wake disorder</th>\n",
       "      <th>Narcolepsy, EDS, Hypersomnia</th>\n",
       "      <th>Catathrenia</th>\n",
       "      <th>control</th>\n",
       "      <th>...</th>\n",
       "      <th>Lowest SaO2 (%)</th>\n",
       "      <th>ODI</th>\n",
       "      <th>90%ODI</th>\n",
       "      <th>EtCO2 (&gt;50mmHg)</th>\n",
       "      <th>Total LMI =PLMI(/h)</th>\n",
       "      <th>Total LM Arousal#</th>\n",
       "      <th>Total LM AI(/h)</th>\n",
       "      <th>PLM Arousal#</th>\n",
       "      <th>PLM AI(/h)</th>\n",
       "      <th>MAI(/h)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>55</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>88.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>48</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>89.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>55</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>86.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>83.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>46.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>89.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 85 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   sex  age  OSA  insomnia  PLMD, RLS  parasomnia  \\\n",
       "0    1   55    1         0          0           0   \n",
       "1    0   48    1         0          0           0   \n",
       "2    0   55    1         0          0           0   \n",
       "3    1   50    1         0          0           0   \n",
       "4    0   57    1         0          0           0   \n",
       "\n",
       "   circadian rhythm sleep-wake disorder  Narcolepsy, EDS, Hypersomnia  \\\n",
       "0                                     0                             0   \n",
       "1                                     0                             0   \n",
       "2                                     0                             0   \n",
       "3                                     0                             0   \n",
       "4                                     0                             0   \n",
       "\n",
       "   Catathrenia  control  ...  Lowest SaO2 (%)  ODI  90%ODI  EtCO2 (>50mmHg)  \\\n",
       "0            0        0  ...             88.0  0.0     0.0              0.0   \n",
       "1            0        0  ...             89.0  0.0     0.0              0.0   \n",
       "2            0        0  ...             86.0  0.0     0.0              0.0   \n",
       "3            0        0  ...             83.0  0.0     0.0              0.0   \n",
       "4            0        0  ...             89.0  0.0     0.0              0.0   \n",
       "\n",
       "   Total LMI =PLMI(/h)  Total LM Arousal#  Total LM AI(/h)  PLM Arousal#  \\\n",
       "0                  3.2                2.0              0.3           0.0   \n",
       "1                  0.0                0.0              0.0           0.0   \n",
       "2                  0.0                0.0              0.0           0.0   \n",
       "3                 46.8                0.0              0.0           0.0   \n",
       "4                  0.0                0.0              0.0           0.0   \n",
       "\n",
       "   PLM AI(/h)  MAI(/h)  \n",
       "0         0.0      0.3  \n",
       "1         0.0      0.0  \n",
       "2         0.0      0.0  \n",
       "3         0.0      0.0  \n",
       "4         0.0      0.0  \n",
       "\n",
       "[5 rows x 85 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_csv('../data/preprocessed_data_a.csv')\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "22f246be",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "target = data.iloc[:, 2:10]\n",
    "features = data.drop(data.iloc[:, 2:10], axis=1)\n",
    "\n",
    "X_train, X_test, y_train, y_test = train_test_split(features, target, test_size=0.3, random_state=42)\n",
    "y_train_converted = y_train.apply(lambda y: y.argmax(), axis=1).values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ac5358fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n"
     ]
    }
   ],
   "source": [
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from skmultilearn.ensemble import RakelD\n",
    "\n",
    "classifier = RakelD(\n",
    "    base_classifier=RandomForestClassifier(),\n",
    "    base_classifier_require_dense=[True, True],\n",
    "    labelset_size=3\n",
    ")\n",
    "\n",
    "classifier.fit(X_train, y_train)\n",
    "y_pred = classifier.predict(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "49fcc097",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py:129: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  return np.array(label_sets)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'base_classifier_require_dense': [False, False], 'labelset_size': 10} 0.826505776449299\n"
     ]
    }
   ],
   "source": [
    "parameters = {\n",
    "    'base_classifier_require_dense': [[True, True], [True, False], [False, False], [False, True]],\n",
    "    'labelset_size': list(range(1, 11))\n",
    "}\n",
    "score = 'f1_samples'\n",
    "\n",
    "clf = GridSearchCV(RakelD(base_classifier=RandomForestClassifier()), parameters, scoring=score)\n",
    "clf.fit(X_train, y_train)\n",
    "\n",
    "print(clf.best_params_, clf.best_score_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d2f795a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 1 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 1 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 1 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 1 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 1 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 2 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 2 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 2 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 2 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 2 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 3 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 3 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 3 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 3 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 3 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 4 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 4 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 4 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 4 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 4 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 5 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 5 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 5 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 5 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 5 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 6 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 6 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 6 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 6 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py\", line 593, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/rakelo.py\", line 121, in fit\n",
      "    return self.classifier.fit(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/problem_transform/br.py\", line 153, in fit\n",
      "    self._generate_partition(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/ensemble/partition.py\", line 140, in _generate_partition\n",
      "    self.partition_ = self.clusterer.fit_predict(X, y)\n",
      "  File \"/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/skmultilearn/cluster/random.py\", line 95, in fit_predict\n",
      "    raise ValueError(\"Cannot include all of {} labels in {} clusters of {} labels\".format(\n",
      "ValueError: Cannot include all of 8 labels in 6 clusters of 1 labels\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n"
     ]
    }
   ],
   "source": [
    "parameters = {\n",
    "    'base_classifier_require_dense': [[True, True], [True, False], [False, False], [False, True]],\n",
    "    'labelset_size': list(range(1, 11)),\n",
    "    'model_count': list(range(1, 11))\n",
    "}\n",
    "score = 'f1_samples'\n",
    "\n",
    "clf = GridSearchCV(RakelO(base_classifier=RandomForestClassifier()), parameters, scoring=score)\n",
    "clf.fit(X_train, y_train)\n",
    "\n",
    "print(clf.best_params_, clf.best_score_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "443db985",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.7244501940491591\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import *\n",
    "\n",
    "accuracy = accuracy_score(y_test, y_pred)\n",
    "print(accuracy)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "73abe456",
   "metadata": {},
   "outputs": [],
   "source": [
    "from skmultilearn.ensemble import RakelO\n",
    "\n",
    "classifier = RakelO(\n",
    "    base_classifier=RandomForestClassifier(),\n",
    "    base_classifier_require_dense=[True, True],\n",
    "    labelset_size=y_train.shape[1] // 4,\n",
    "    model_count=6\n",
    ")\n",
    "\n",
    "classifier.fit(X_train, y_train)\n",
    "y_pred = classifier.predict(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "623cba39",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.7018111254851229\n"
     ]
    }
   ],
   "source": [
    "accuracy = accuracy_score(y_test, y_pred)\n",
    "print(accuracy)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7f73d58a",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=1 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=1 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=1 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=1 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=1 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=2 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=2 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=2 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=2 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=2 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=3 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=3 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=3 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=3 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=3 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=4 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=4 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=4 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=4 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=4 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=5 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=5 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=5 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=5 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=5 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=6 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=6 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=6 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=6 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=6 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=7 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=7 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=7 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=7 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=7 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=8 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=8 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=8 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=8 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=8 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=9 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=9 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=9 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=9 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=9 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'k': 6} 0.021388338701617115\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/hyewonhan/miniconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass n_neighbors=6 as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n"
     ]
    }
   ],
   "source": [
    "from skmultilearn.adapt import BRkNNbClassifier\n",
    "from sklearn.model_selection import GridSearchCV\n",
    "\n",
    "X = X_train.to_numpy()\n",
    "y = y_train.to_numpy()\n",
    "\n",
    "parameters = {'k': range(1,10)}\n",
    "score = 'f1_macro'\n",
    "\n",
    "clf = GridSearchCV(BRkNNbClassifier(), parameters, scoring=score)\n",
    "clf.fit(X, y)\n",
    "\n",
    "print(clf.best_params_, clf.best_score_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "97cfd61a",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'c_k': 0.03125} 0.15509160354531726\n"
     ]
    }
   ],
   "source": [
    "from skmultilearn.adapt import MLTSVM\n",
    "\n",
    "parameters = {'c_k': [2**i for i in range(-5, 5, 2)]}\n",
    "score = 'f1_macro'\n",
    "\n",
    "X_matrix = np.asmatrix(X)\n",
    "y_matrix = np.asmatrix(y)\n",
    "\n",
    "clf = GridSearchCV(MLTSVM(), parameters, scoring=score)\n",
    "clf.fit(X_matrix, y_matrix)\n",
    "\n",
    "print(clf.best_params_, clf.best_score_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "08a3b4c8",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "0",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/core/indexes/base.py:3621\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3620\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3621\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3622\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/_libs/index.pyx:136\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/_libs/index.pyx:163\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5198\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5206\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 0",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Input \u001b[0;32mIn [22]\u001b[0m, in \u001b[0;36m<cell line: 4>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mskmultilearn\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01madapt\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m MLARAM\n\u001b[1;32m      3\u001b[0m classifier \u001b[38;5;241m=\u001b[39m MLARAM(threshold\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.05\u001b[39m, vigilance\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.95\u001b[39m)\n\u001b[0;32m----> 4\u001b[0m \u001b[43mclassifier\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit\u001b[49m\u001b[43m(\u001b[49m\u001b[43mX_train\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43my_train\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/skmultilearn/adapt/mlaram.py:164\u001b[0m, in \u001b[0;36mMLARAM.fit\u001b[0;34m(self, X, y)\u001b[0m\n\u001b[1;32m    162\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(y, numpy\u001b[38;5;241m.\u001b[39mmatrix):\n\u001b[1;32m    163\u001b[0m     y \u001b[38;5;241m=\u001b[39m numpy\u001b[38;5;241m.\u001b[39masarray(y)\n\u001b[0;32m--> 164\u001b[0m is_more_dimensional \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mint\u001b[39m(\u001b[38;5;28mlen\u001b[39m(\u001b[43mX\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39mshape) \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m1\u001b[39m)\n\u001b[1;32m    165\u001b[0m X \u001b[38;5;241m=\u001b[39m _normalize_input_space(X)\n\u001b[1;32m    167\u001b[0m y_0 \u001b[38;5;241m=\u001b[39m _get_label_vector(y, \u001b[38;5;241m0\u001b[39m)\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/core/frame.py:3505\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3503\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mnlevels \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m   3504\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[0;32m-> 3505\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3506\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[1;32m   3507\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m [indexer]\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/core/indexes/base.py:3623\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3621\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine\u001b[38;5;241m.\u001b[39mget_loc(casted_key)\n\u001b[1;32m   3622\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[0;32m-> 3623\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3624\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3625\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3626\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3627\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3628\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 0"
     ]
    }
   ],
   "source": [
    "from skmultilearn.adapt import MLARAM\n",
    "\n",
    "classifier = MLARAM(threshold=0.05, vigilance=0.95)\n",
    "classifier.fit(X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62f82f50",
   "metadata": {},
   "outputs": [],
   "source": [
    "prediction = classifier.predict(X_test)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
